http-server-netty/src/main/java/io/micronaut/http/server/netty/handler/accesslog/HttpAccessLogHandler.java

```diff
@@ -156,32 +149,105 @@ private void log(ChannelHandlerContext ctx, Object msg, ChannelPromise promise,
-            ctx.write(response, promise);
-            return true;
+    private void processWriteEvent(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        AccessLogHolder accessLogHolder = getAccessLogHolder(ctx, false);
+        if (accessLogHolder != null) {
+            boolean isContinueResponse = msg instanceof HttpResponse && ((HttpResponse) msg).status().equals(HttpResponseStatus.CONTINUE);
+            AccessLog accessLogger = accessLogHolder.getLogForResponse(
+                    msg instanceof HttpMessage ? (HttpMessage) msg : null,
+                    msg instanceof LastHttpContent && !isContinueResponse);
+            if (accessLogger != null && !isContinueResponse) {
+                if (msg instanceof HttpResponse) {
+                    accessLogger.onResponseHeaders(ctx, ((HttpResponse) msg).headers(), ((HttpResponse) msg).status().codeAsText().toString());
+                }
+                if (msg instanceof LastHttpContent) {
+                    accessLogger.onLastResponseWrite(((LastHttpContent) msg).content().readableBytes());
+                    log(ctx, msg, promise, accessLogger);
+                    return;
+                } else if (msg instanceof ByteBufHolder) {
+                    accessLogger.onResponseWrite(((ByteBufHolder) msg).content().readableBytes());
+                } else if (msg instanceof ByteBuf) {
+                    accessLogger.onResponseWrite(((ByteBuf) msg).readableBytes());
+                }
+            }
+        }
+        super.write(ctx, msg, promise);
+    }
+
+    @Nullable
+    @Contract("_, true -> !null") // can only return null when createIfMissing is false
+    private AccessLogHolder getAccessLogHolder(ChannelHandlerContext ctx, boolean createIfMissing) {
+        final Attribute<AccessLogHolder> attr = ctx.channel().attr(ACCESS_LOGGER);
+        AccessLogHolder holder = attr.get();
+        if (holder == null) {
+            if (!createIfMissing) {
+                return null;
+            }
+            holder = new AccessLogHolder();
+            attr.set(holder);
         }
-        accessLogger.onResponseHeaders(ctx, response.headers(), status.codeAsText().toString());
-        return false;
+        return holder;
     }
 
-    private void processWriteEvent(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
-        final AccessLog accessLogger = ctx.channel().attr(ACCESS_LOGGER).get();
-        if (accessLogger != null) {
-            if (msg instanceof HttpResponse && processHttpResponse((HttpResponse) msg, accessLogger, ctx, promise)) {
-                return;
+    /**
+     * Holder for {@link AccessLog} instances. {@link AccessLog} can only handle one concurrent request at a time, this
+     * class multiplexes access where necessary.
+     */
+    private final class AccessLogHolder {
+        private final Map<Long, AccessLog> liveAccessLogsByStreamId = new HashMap<>();
```

Comment:
can this be concurrently accessed?

Reply:
no, the holder is local to the channel, and the handler is only called on the event loop